File: src/test/java/functor/FunctorTest.java
package functor;

import org.junit.jupiter.api.Test;

import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.assertEquals;

class FunctorTest {

    @Test
    public void fmapOnNothing_IsVoid() {
        assertEquals(new Nothing(), new Nothing().fmap(plus10()));
    }

    @Test
    public void fmapOnJust_MapsFunctionToInternalValue() {
        assertEquals(new Just(15), new Just(5).fmap(plus10()));
    }

    private Function<Integer, Integer> plus10() {
        return value -> value + 10;
    }
}

File: src/test/kotlin/root_k/functor/FunctorTest.kt
package root_k.functor

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.fail
import org.junit.jupiter.api.Test
import root_k.functor.Maybe.Just
import root_k.functor.Maybe.None

class FunctorTest {

    @Test
    fun `flat mapping nothing has no effect`() {
        assertEquals(None, None fmap { fail<Unit>("should not be applied")})
    }

    @Test
    fun `fmap applies function on Just`() {
        assertEquals(Just(15), Just(5) fmap {it + 10})
    }
}